### PR TITLE
Conditional steps based on HTTP context

### PR DESCRIPTION
**Pull request by emswbandara**

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsCookie.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        this.wrapped = cookie;
+    }
+
+    @Override
+    public Object getMember(String name) {
+        if (wrapped == null) {
+            return super.getMember(name);
+        }
+        switch (name) {
+            case "name":
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsCookie.java]: Shall we add constants to these strings?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/ConditionalRequestCoordinator.java [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+/**
+ * Conditional Request Coordinator.
+ */
+public class ConditionalRequestCoordinator extends AbstractRequestCoordinator implements RequestCoordinator {
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/ConditionalRequestCoordinator.java]: Shall we extend DefaultRequestCoordinator and override the required methods?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/ConditionalRequestCoordinator.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    private AuthenticationRequestCacheEntry getAuthenticationRequest(HttpServletRequest request,
+                                                                     String sessionDataKey) {
+
+        AuthenticationRequestCacheEntry authRequest = getAuthenticationRequestFromRequest(request);
+        if (authRequest == null) {
+            authRequest = FrameworkUtils.getAuthenticationRequestFromCache(sessionDataKey);
+        }
+        return authRequest;
+    }
+}
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/ConditionalRequestCoordinator.java]: Add new line at the end of the file

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsServletResponse.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Javascript wrapper for Java level HttpServletResponse.
+ * This provides controlled access to HttpServletResponse object via provided javascript native syntax.
+ * e.g
+ * response.headers.["Set-Cookie"] = ['crsftoken=xxxxxssometokenxxxxx']
```
Comment (ruwanta) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/js/JsServletResponse.java]: We need to have better cookie add
e.g. 
var newCookie = {'name': 'myName'};
newCookie.path = '/foo/bar';
response.cookies.push(newCookie);

**Pull request merged**